### PR TITLE
Implement visitor plugin

### PR DESCRIPTION
**Pull request by ForNeVeR**
It will close issue #136. Probably there's still more work with user role and affiliation detection. We will need to hack Smack to sort that out, I'll create a new task later (done: #270).

### CODE REVIEW

**Comment by ForNeVeR** [comment]
cc @rexim

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala [hunk 1]
```diff
@@ -5,11 +5,26 @@ import akka.actor.ActorRef
 import ru.org.codingteam.horta.messages._
 
 class MucParticipantStatusListener(muc: MultiUserChat, room: ActorRef) extends DefaultParticipantStatusListener {
+
+  // TODO: Test for changing the user status from visitor to participant and back.
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: Should we create an issue for this? You know, with `refactoring` tag or something like that.

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala [hunk 2]
```diff
@@ -5,11 +5,26 @@ import akka.actor.ActorRef
...
+      case "owner" => Owner
+      case "admin" => Admin
+      case _ => User
+    }
 
-    room ! UserJoined(participant, affiliation)
+    val role = roleName match {
+      case "moderator" => Moderator
+      case "participant" => Participant
+      case _ => Visitor // TODO: Check the real value for visitor. Currently I have no time to experiment. ~ F
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: Does this need a separate issue as well?

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucMessageHandler.scala [hunk 3]
```diff
@@ -24,7 +24,8 @@ class MucMessageHandler(locale: LocaleDefinition,
 
   val core = context.actorSelection("/user/core")
 
-  var participants = Map[String, Affinity](s"$roomJID/$nickname" -> User)
+  val participant = Protocol.Participant(s"$roomJID/$nickname", User, Participant) // TODO: Get own role and affiliation
```
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucMessageHandler.scala]: Could you please briefly explain what this TODO means?

Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucMessageHandler.scala]: It means that currently the bot always thinks that he is `User` and `Participant` in any room he can enter. It is not always true, sometimes he can be `Moderator` or even `Owner`.

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala (hunk 1 above)
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: I have already created it; #270 is precisely for that.

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala (hunk 2 above)
Comment (ForNeVeR) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: Hm, maybe you're right. I'll create it.

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala (hunk 1 above)
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: ok

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala (hunk 2 above)
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucParticipantStatusListener.scala]: ok

File: src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucMessageHandler.scala (hunk 3 above)
Comment (rexim) [inline src/main/scala/ru/org/codingteam/horta/protocol/jabber/MucMessageHandler.scala]: Ok, got it

**Comment by rexim** [comment]
Approved

**Pull request merged**